pkg/manager/impl/executions/quality_of_service.go

```diff
@@ -0,0 +1,182 @@
+		if input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.GetSpec() != nil {
+			duration, err := ptypes.Duration(input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.
+				GetSpec().QueueingBudget)
+			if err != nil {
+				return QualityOfServiceSpec{}, errors.NewFlyteAdminErrorf(codes.InvalidArgument,
+					"Invalid custom quality of service set in workflow [%v], failed to parse duration [%v] with: %v",
+					workflowIdentifier,
+					input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec().QueueingBudget, err)
+			}
+			return QualityOfServiceSpec{
+				QueuingBudget: duration,
+			}, nil
+		}
+		qualityOfServiceTier = input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.GetTier()
+	}
+
+	// If nothing in the hierarchy of registrable entities has set the quality of service,
+	// see if an override exists in the matchable attributes resource table.
+	if qualityOfServiceTier == core.QualityOfService_UNDEFINED {
+		qualityOfService, err := q.getQualityOfServiceFromDb(ctx, workflowIdentifier)
+		if err != nil {
+			return QualityOfServiceSpec{}, err
+		}
+		if qualityOfService != nil && qualityOfService.GetSpec() != nil {
+			duration, err := ptypes.Duration(qualityOfService.GetSpec().QueueingBudget)
+			if err != nil {
+				return QualityOfServiceSpec{}, errors.NewFlyteAdminErrorf(codes.InvalidArgument,
+					"Invalid custom quality of service set in overridable matching attributes for [%v],"+
+						"failed to parse duration [%v] with: %v", workflowIdentifier,
+					input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec().QueueingBudget, err)
+			}
+			return QualityOfServiceSpec{
+				QueuingBudget: duration,
+			}, nil
+		} else if qualityOfService != nil && qualityOfService.GetTier() != core.QualityOfService_UNDEFINED {
+			qualityOfServiceTier = input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.GetTier()
+		}
+	}
+
+	// If we've come all this way and at no layer is an overridable configuration for the quality of service tier
+	// set, use the default values from the admin application config.
+	if qualityOfServiceTier == core.QualityOfService_UNDEFINED {
+		var ok bool
+		qualityOfServiceTier, ok = q.config.QualityOfServiceConfiguration().GetDefaultTiers()[input.ExecutionCreateRequest.Domain]
+		if !ok {
+			// No queueing budget to set when no default is specified
+			return QualityOfServiceSpec{}, nil
+		}
+	}
+	executionValues, ok := q.config.QualityOfServiceConfiguration().GetTierExecutionValues()[qualityOfServiceTier]
+	if !ok {
+		// No queueing budget to set when no default is specified
+		return QualityOfServiceSpec{}, nil
+	}
+	// Config values should always be vetted so there's no need to check the error from conversion.
+	duration, _ := ptypes.Duration(executionValues.QueueingBudget)
+
+	return QualityOfServiceSpec{
```

Comment:
can we add a debug log line to each case so we can more easily understand what was used?

Reply:
done